integration_tests/src/main/python/url_test.py

```diff
@@ -0,0 +1,234 @@
+    "http://user:pass@host",
+    "http://user:pass@host/",
+    "http://user:pass@host/?#",
+    "http://user:pass@host/file;param?query;p2",
+    "inva lid://user:pass@host/file;param?query;p2",
+    # "http://[1:2:3:4:5:6:7:8]",
+    # "http://[1::]",
+    # "http://[1:2:3:4:5:6:7::]",
+    # "http://[1::8]",
+    # "http://[1:2:3:4:5:6::8]",
+    # "http://[1:2:3:4:5:6::8]",
+    # "http://[1::7:8]",
+    # "http://[1:2:3:4:5::7:8]",
+    # "http://[1:2:3:4:5::8]",
+    # "http://[1::6:7:8]",
+    # "http://[1:2:3:4::6:7:8]",
+    # "http://[1:2:3:4::8]",
+    # "http://[1::5:6:7:8]",
+    # "http://[1:2:3::5:6:7:8]",
+    # "http://[1:2:3::8]",
+    # "http://[1::4:5:6:7:8]",
+    # "http://[1:2::4:5:6:7:8]",
+    # "http://[1:2::8]",
+    # "http://[1::3:4:5:6:7:8]",
+    # "http://[1::3:4:5:6:7:8]",
+    # "http://[1::8]",
+    # "http://[::2:3:4:5:6:7:8]",
+    # "http://[::2:3:4:5:6:7:8]",
+    # "http://[::8]",
+    # "http://[::]",
+    # "http://[fe80::7:8%eth0]",
+    # "http://[fe80::7:8%1]",
+    # "http://[::255.255.255.255]",
+    # "http://[::ffff:255.255.255.255]",
+    # "http://[::ffff:0:255.255.255.255]",
+    # "http://[2001:db8:3:4::192.0.2.33]",
+    # "http://[64:ff9b::192.0.2.33]"
+]
+
+edge_cases_gen = SetValuesGen(StringType(), edge_cases)
+
+url_gen = StringGen(url_pattern)
+    
+def test_parse_url_protocol():
+    assert_gpu_and_cpu_are_equal_collect(
+            lambda spark : unary_op_df(spark, url_gen).selectExpr(
+                "a",
+                "parse_url(a, 'PROTOCOL')"
+                ))
+    
+def test_parse_url_protocol_edge_cases():
+    assert_gpu_and_cpu_are_equal_collect(
+            lambda spark : unary_op_df(spark, edge_cases_gen).selectExpr(
+                "a",
+                "parse_url(a, 'PROTOCOL')"
+                ))
+
+@allow_non_gpu('ProjectExec', 'ParseUrl')
+def test_parse_url_host_fallback():
+    assert_gpu_fallback_collect(
+            lambda spark : unary_op_df(spark, url_gen).selectExpr(
+                "a",
+                "parse_url(a, 'HOST')"
+                ),
+            'ParseUrl')
+    
+@allow_non_gpu('ProjectExec', 'ParseUrl')
+def test_parse_url_path_fallback():
+    assert_gpu_fallback_collect(
+            lambda spark : unary_op_df(spark, url_gen).selectExpr(
+                "a",
+                "parse_url(a, 'PATH')"
+                ),
+            'ParseUrl')
+
+@allow_non_gpu('ProjectExec', 'ParseUrl')
+def test_parse_url_query_fallback():
+    assert_gpu_fallback_collect(
+            lambda spark : unary_op_df(spark, url_gen).selectExpr(
+                "a",
+                "parse_url(a, 'QUERY')"
+                ),
+            'ParseUrl')
+
+@allow_non_gpu('ProjectExec', 'ParseUrl')   
+def test_parse_url_ref_fallback():
+    assert_gpu_fallback_collect(
+            lambda spark : unary_op_df(spark, url_gen).selectExpr(
+                "a",
+                "parse_url(a, 'REF')"
+                ),
+            'ParseUrl')
```

Comment:
all of these should be refactored to a single test parameterized on the literal rather than duplicated.

Reply:
done.